File: hw/vizcaino/u3/hw22SOLIDMaze/MazeSystem/src/main/java/ec/edu/espe/mazesystem/model/AbstractMaze.java
package ec.edu.espe.mazesystem.model;  

import java.util.ArrayList;
import java.util.List;

public abstract class AbstractMaze {
    protected int width;
    protected int height;
    protected List<List<List<Room>>> rooms;
    protected Path solutionPath;

    public AbstractMaze(int width, int height) {
        this.width = width;
        this.height = height;
        this.rooms = new ArrayList<>();
        this.solutionPath = new Path();
        initializeRooms();
    }

    public AbstractMaze() {
        this(5, 5);
    }

    private void initializeRooms() {
        for (int i = 0; i < height; i++) {
            List<List<Room>> row = new ArrayList<>();
            for (int j = 0; j < width; j++) {
                List<Room> roomList = new ArrayList<>();
                row.add(roomList);
            }
            rooms.add(row);
        }
    }

    public abstract void generate();

    public Room getRoom(int x, int y) {
        if (x >= 0 && x < width && y >= 0 && y < height) {
            List<Room> roomList = rooms.get(y).get(x);
            return roomList.isEmpty() ? null : roomList.get(0);
        }
        return null;
    }

    public void addRoom(int x, int y, Room room) {
        if (x >= 0 && x < width && y >= 0 && y < height) {
            rooms.get(y).get(x).add(room);
        }
    }

    public Path getSolutionPath() {
        return solutionPath;
    }

    public void setSolutionPath(Path solutionPath) {
        this.solutionPath = solutionPath;
    }

    public int getWidth() {
        return width;
    }

    public int getHeight() {
        return height;
    }
}

File: hw/chavez/u3/hw22SOLIDMaze/hw22Python/model/abstract_maze.py
from abc import ABC, abstractmethod

from model.path import Path


class AbstractMaze(ABC):

    def __init__(self, width=5, height=5):
        self.width = width
        self.height = height
        self.rooms = []
        self.solution_path = Path()

        self._initialize_rooms()

    def _initialize_rooms(self):
        for _ in range(self.height):
            row = []

            for _ in range(self.width):
                room_list = []
                row.append(room_list)

            self.rooms.append(row)

    @abstractmethod
    def generate(self):
        pass

    def get_room(self, x, y):
        if 0 <= x < self.width and 0 <= y < self.height:
            room_list = self.rooms[y][x]

            if len(room_list) == 0:
                return None

            return room_list[0]

        return None

    def add_room(self, x, y, room):
        if 0 <= x < self.width and 0 <= y < self.height:
            self.rooms[y][x].append(room)

    def get_solution_path(self):
        return self.solution_path

    def set_solution_path(self, solution_path):
        self.solution_path = solution_path

    def get_width(self):
        return self.width

    def get_height(self):
        return self.height